Sources/Async/EventLoop/Kqueue/KqueueEventLoop.swift

```diff
@@ -41,89 +32,80 @@ public final class KqueueEventLoop: EventLoop {
-        writeSources = .init(start: .allocate(capacity: maxDescriptor), count: maxDescriptor)
-
-        tasks = []
+        /// set async task to nil
+        task = nil
     }
 
     /// See EventLoop.onReadable
     public func onReadable(descriptor: Int32, _ callback: @escaping EventLoop.EventCallback) -> EventSource {
-        let source = KqueueEventSource(descriptor: descriptor, kq: kq, callback: callback)
-        source.event.filter = Int16(EVFILT_READ)
-        readSources[Int(descriptor)] = source
-        return source
+        return KqueueEventSource(descriptor: descriptor, kq: kq, type: .read, callback: callback)
     }
 
     /// See EventLoop.onWritable
     public func onWritable(descriptor: Int32, _ callback: @escaping EventLoop.EventCallback) -> EventSource {
-        let source = KqueueEventSource(descriptor: descriptor, kq: kq, callback: callback)
-        source.event.filter = Int16(EVFILT_WRITE)
-        writeSources[Int(descriptor)] = source
-        return source
+        return KqueueEventSource(descriptor: descriptor, kq: kq, type: .write, callback: callback)
+    }
+
+
+    /// See EventLoop.ononTimeout
+    public func onTimeout(timeout: Int, _ callback: @escaping EventLoop.EventCallback) -> EventSource {
+        return KqueueEventSource(descriptor: 1, kq: kq, type: .timer(timeout: timeout), callback: callback)
     }
 
     /// See EventLoop.async
     public func async(_ callback: @escaping EventLoop.AsyncCallback) {
-        tasks.append(callback)
+        if task != nil {
+            // if there is a task waiting, run the event loop
+            // until it has been executed
+            run()
+        }
+
+        /// set the new task
+        task = callback
     }
 
     /// See EventLoop.run
     public func run() {
-        // run all async tasks
-        if self.tasks.count > 0 {
-            let tasks = self.tasks
-            self.tasks = []
-            for task in tasks {
-                task()
-            }
+        // while tasks are available, run them
+        while let task = self.task {
+            // make sure to set this task to nil, so that
+            // the task can set a new one if it needs
+            self.task = nil
+
+            // run the task, potentially creating a new task
+            // in the process
+            task()
         }
 
-        var timeout = timespec()
-        timeout.tv_sec = 0
-        timeout.tv_nsec = 100_000_000 // 0.1 seconds
-
         // check for new events
-        let eventCount = kevent(kq, nil, 0, eventlist.baseAddress, Int32(eventlist.count), &timeout)
+        let eventCount = kevent(kq, nil, 0, eventlist.baseAddress, Int32(eventlist.count), nil)
         guard eventCount >= 0 else {
-            print("An error occured while running kevent: \(eventCount).")
-            return
+            switch errno {
+            case EINTR:
+                run() // run again
```

Comment:
Is this really the best way to handle interrupts?

Reply:
> When kevent() call	fails with EINTR error,	all changes in the changelist
     have been applied.

From this I believe so.